Fix ScrollContentPresenter arrange when ScrollViewer.HorizontalScrollBarVisibility is Disabled.

grokys opened the pull request:
When `ScrollViewer.HorizontalScrollBarVisibility` is `Disabled`, then we should not arrange the content to be wider than the presenter width.

Fixes #1065

grokys committed:
Added failing test for #1065.

Also add setter for `ScrollContentPresenter.CanScrollHorizontally`.

grokys committed:
Respect CanScrollHorizontally in Arrange.

Fixes #1065

grokys committed:
Fix layout in ScrollContentPresenter.

- Share common layout logic between `Border`, `ContentPresenter` and `ScrollContentPresenter`
- Added a bunch of tests for things not previously convered
- Fix `ScrollContentPresenter` child layout

grokys commented:
From @MonkAlex in gitter:

elements correct start on new line, but has horizontal scrollbar on start - and hide then after resize

grokys commented:
@MonkAlex just pushed more fixes, could you try now?

MonkAlex commented:
More bugged now.

Horizontal scrollbar hide now, but after start app show only first 40 elements, after resize - all 100.

Vertical scrollbar work for 40 (after resize all work), with keyboard can move to any element, but its not drawing.
![horizontal](https://user-images.githubusercontent.com/6784764/34880883-4d58a832-f7cb-11e7-8a72-e7a8ad75ca79.gif)

grokys committed:
Merge branch 'master' into fixes/1065-canscrollhorizontally

 Conflicts:
	src/Avalonia.Controls/Presenters/ScrollContentPresenter.cs

grokys committed:
Fix window child constraint.

Instead of measuring the child and then clipping the child's desired size based on `SizeToContent`, instead pass a constraint based on `SizeToContent` to the child.

grokys commented:
@MonkAlex I've committed some more changes - could you see if that fixes things for you? If no, would it be possible to get a hold of your code so I can debug it there?

grokys committed:
Fixed malformed doc comment.

MonkAlex commented:
So, now xaml parse is failed, no idea - why.
I attach project. Work (bad) on 4416 build, but fail on current branch.

[AvaloniaList_0.5.2-build4416-alpha.zip](https://github.com/AvaloniaUI/Avalonia/files/1664811/AvaloniaList_0.5.2-build4416-alpha.zip)

Maybe u has simple xaml for it - just want "library" with labeled images

grokys commented:
@MonkAlex the reason for the XAML error was that `ScrollViewer.CanScrollHorizontally` was replaced by `HorizontalScrollBarVisibility="Disabled"` in #1348.

However I'm still seeing a layout problem in your test project that wasn't showing up in my testing, trying to fix it now!

grokys commented:
Actually, it seems to be working now!

MonkAlex commented:
Work as expected, thanks.

grokys committed:
Merge branch 'master' into fixes/1065-canscrollhorizontally

jkoritzinsky commented on src/Avalonia.Controls/Presenters/ContentPresenter.cs:
```diff
@@ -340,94 +338,121 @@ protected virtual IControl CreateChild()
         {
-            var child = Child;
-            var padding = Padding + new Thickness(BorderThickness);
+            return Border.MeasureOverrideImpl(availableSize, Child, Padding, BorderThickness);
```
Any particular reason that `MeasureOverrideImpl` is defined on `Border` but `ArrangeOverrideImpl` is defined on `ContentPresenter`? I feel that it would be more consistent if both were defined on `ContentPresenter`.

grokys replied:
Yeah, the reason is that the `MeasureOverride` algorithm applies to `Border` and `ContentPresenter` and the `ArrangeOverride` algorithm applies to `ContentPresenter` and `ScrollContentPresenter`. To me `Border` is conceptually a "lower" level control than `ContentPresenter`, so `Border` calling out to `ContentPresenter` feels wrong. Does that make sense?

jkoritzinsky replied:
Makes sense.

jkoritzinsky reviewed (approved):
I've added a small comment, but other than that LGTM!
